### PR TITLE
Feat/8762 perpetual markets support in existing apis

### LINKED ISSUE #8120: Datanode API for spot market/data/governance
# Feature Overview

In order to access spots through datanode, the datanode model needs to be updated to represent spot market, market data, LP commitment, governance proposal. 

# Acceptance Criteria
TBD

### LINKED ISSUE #8762: [Feature]: Datanode API: Perpetual markets can be seen through all existing API that return a market
# Feature Overview

**In order to** Ensure that perpetual markets are visible and return the correct data across all existing datanode APIs
**We will** perform an code-review/audit/add unittests
**So that** perpetuals are as available and usable through our API as fixed expiry futures.

Given that we store the market definition in postgres as a jsonb I think this will just work. This ticket is sanity check this, and to write some unit-tests to confirm this to be true. This include new additions to any data-sourcing. We might also want to consider what new fields we could add to the market-data, such as funding rates.

We should also have a conversation with FE about how they expect to be able to filter between market types. Console currently have a TAB that toggles between product types. We should understand what they expect from our API's to support this:
image

## Specs
- [Link](https://github.com/vegaprotocol/specs/blob/cosmicelevator/protocol/0048-DSRI-data_source_internal.md) to spec or section within a spec

## Tasks
A checklist of the tasks that are needed to develop the feature and meet the acceptance criteria and feature test scenarios.
- [ ]
- [ ]

## Feature test scenarios
Detailed scenarios that can be executed as feature tests to verify that the feature has been implemented as expected.

GIVEN (setup/context) 
WHEN (action) 
THEN (assertion) For example...
See [here](https://github.com/vegaprotocol/vega/tree/develop/core/integration) for more format information and examples.

### Additional Details (optional)
Any additional information that provides context or gives information that will help us develop the feature.

### LINKED ISSUE #8951: GraphQL objects for Instrument and InstrumentConfiguration does not support Spots and Perps
The definition for Instrument and InstrumentConfiguration that used in Market and Proposal objects have not been updated to include the Spot and Perps instrument types. The definition will need to be changed for these GQL types as such may cause some breaking change and will require deprecation, for example:

```
type InstrumentConfiguration {
  "Full and fairly descriptive name for the instrument"
  name: String!
  "A short non necessarily unique code used to easily describe the instrument (e.g: FX:BTCUSD/DEC18)"
  code: String!
  "Future product specification"
  futureProduct: FutureProduct
}
```

will need to become:

```
type InstrumentConfiguration {
  "Full and fairly descriptive name for the instrument"
  name: String!
  "A short non necessarily unique code used to easily describe the instrument (e.g: FX:BTCUSD/DEC18)"
  code: String!
  "Future product specification"
  futureProduct: FutureProduct @deprecated(reason: "Use product union instead"
  "Product specification"
  product: Product
}

union Product = FutureProduct | SpotProduct | PerpetualProduct

type SpotProduct {
    ...
}

type PerpetualProduct {
    ...
}

```

### PR DESCRIPTION
**Pull request by guoguojin**
Closes #8762 
Closes #8120 
Closes #8951

### CODE REVIEW

File: datanode/gateway/graphql/schema.graphql [hunk 1]
```diff
@@ -1509,6 +1509,57 @@ type Future {
   dataSourceSpecBinding: DataSourceSpecToFutureBinding!
 }
 
+"A spot FX product"
+type Spot {
+  "The underlying base asset for the spot product"
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:1514]:
```suggestion
  "Underlying base asset for the spot product"
```

File: datanode/gateway/graphql/schema.graphql [hunk 2]
```diff
@@ -1509,6 +1509,57 @@ type Future {
...
+  baseAsset: Asset!
+  "The underlying quote asset for the spot product"
+  quoteAsset: Asset!
+  "Name of the instrument"
+  name: String!
+}
+
+"A perpetual future product"
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:1522]:
```suggestion
"Perpetual future product"
```

File: datanode/gateway/graphql/schema.graphql [hunk 3]
```diff
@@ -3822,6 +3928,62 @@ type UpdateMarketState {
...
+  decimal_places: Int!
+  "Optional spot market metadata tags"
+  metadata: [String!]!
+  "Price monitoring parameters"
+  priceMonitoringParameters: PriceMonitoringParameters!
+  "Specifies parameters related to target stake calculation"
+  targetStakeParameters: TargetStakeParameters!
+  "New spot market risk model parameters"
+  riskParameters: RiskModel
+  "Decimal placesfor order sizes, sets what size the smallest order / position on the spot market can be"
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:3945]:
```suggestion
  "Decimal places for order sizes, sets what size the smallest order / position on the spot market can be"
```

File: datanode/gateway/graphql/schema.graphql [hunk 4]
```diff
@@ -3822,6 +3928,62 @@ type UpdateMarketState {
...
+  liquiditySLAParams: LiquiditySLAParameters!
+}
+
+type LiquiditySLAParameters {
+  priceRange: String!
+  "Specifies the minimum fraction of time LPs must spend 'on the book' providing their committed liquidity"
+  commitmentMinTimeFraction: String!
+  "Specifies how often the quality of liquidity supplied by the LPS is evaluated and fees arising from that period are earmarked for specific parties"
+  providersFeeCalculationTimeStep: Int!
+  "Specified the number of liquidity epochs over which past performance will continue to affect rewards"
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:3978]:
```suggestion
  "Specifies the number of liquidity epochs over which past performance will continue to affect rewards"
```

**Review by candida-d — DISMISSED** [review]

**Review by candida-d — DISMISSED** [review]

**Review by wwestgarth — DISMISSED** [review]

**Review by candida-d — DISMISSED** [review]

**Comment by wwestgarth** [comment]
Sorry I only just spotted this, feel free to spin it off as a new ticket/PR:
Can we also add the new internal-time-trigger `DataSourceSpecConfigurationTimeTrigger` data source to this union:
```
union InternalDataSourceKind = DataSourceSpecConfigurationTime
```

In the protos it the internal data sources look like this:
```
// Internal data source used for emitting timestamps.
message DataSourceSpecConfigurationTime {
  // Conditions that the timestamps should meet in order to be considered.
  repeated vega.data.v1.Condition conditions = 1;
}

// Internal data source used for emitting timestamps automatically using predefined intervals and conditions.
message DataSourceSpecConfigurationTimeTrigger {
  // Conditions that the timestamps need to meet in order to be considered.
  repeated vega.data.v1.Condition conditions = 1;
  // An internal time trigger
  repeated vega.data.v1.InternalTimeTrigger triggers = 2;
}

// Top level object used for all internal data sources.
// It contains one of any of the defined source type variants.
message DataSourceDefinitionInternal {
  // Types of internal data sources
  oneof source_type {
    DataSourceSpecConfigurationTime time = 1;
    DataSourceSpecConfigurationTimeTrigger time_trigger = 2;
  }
}
```

**Review by wwestgarth — DISMISSED** [review]

**Review by wwestgarth — DISMISSED** [review]

**Review by candida-d — DISMISSED** [review]

File: datanode/gateway/graphql/schema.graphql [hunk 5]
```diff
@@ -3864,6 +4026,9 @@ union ProposalChange =
   | NewTransfer
   | CancelTransfer
   | UpdateMarketState
```
Comment (edd) [inline datanode/gateway/graphql/schema.graphql:4028]: **question (non-blocking)**:  is a new perpetual market just a `NewMarket`, or like spot is it getting a `NewSpotMarket`? 

It seems odd that `NewSpotMarket` is separate - like it's only reasonable for historical reasons that it is separate. I'd prefer it to be merged in to `NewMarket`. But it's not a blocker.

Comment (guoguojin) [reply datanode/gateway/graphql/schema.graphql:4028]: I'm just going by what is in the protos @edd, I think you're right though, I think it should be a configuration for a NewMarket request, perhaps raise it in a separate ticket as it has an impact on core

**Review by wwestgarth — APPROVED** [review]

**Review by mattrussell36 — APPROVED** [review]

**Review by candida-d — APPROVED** [review]

**Review by jeremyletang — APPROVED** [review]

**Pull request merged**